tests/Transport/HttpTransportTest.php

```diff
@@ -250,6 +251,80 @@ public function testSendReturnsRejectedPromiseIfSendingFailedDueToHttpClientExce
         $this->assertSame($event, $promiseResult->getEvent());
     }
 
+    /**
+     * @group time-sensitive
+     */
+    public function testSendReturnsRejectedPromiseIfExceedingRateLimits(): void
+    {
+        ClockMock::withClockMock(1644105600);
+
+        $dsn = Dsn::createFromString('http://public@example.com/sentry/1');
+        $event = Event::createEvent();
+
+        /** @var LoggerInterface&MockObject $logger */
+        $logger = $this->createMock(LoggerInterface::class);
+        $logger->expects($this->exactly(2))
+            ->method('warning')
+            ->withConsecutive(
+                ['Rate limited exceeded for requests of type "event", backing off until "2022-02-06T00:01:00+00:00".', ['event' => $event]],
+                ['Rate limit exceeded for sending requests of type "event".', ['event' => $event]]
+            );
+
+        $this->payloadSerializer->expects($this->once())
+            ->method('serialize')
+            ->with($event)
+            ->willReturn('{"foo":"bar"}');
+
+        $this->requestFactory->expects($this->once())
+            ->method('createRequest')
+            ->with('POST', $dsn->getStoreApiEndpointUrl())
+            ->willReturn(new Request('POST', 'http://www.example.com'));
+
+        $this->streamFactory->expects($this->once())
+            ->method('createStream')
+            ->with('{"foo":"bar"}')
+            ->willReturnCallback([Utils::class, 'streamFor']);
+
+        $this->httpClient->expects($this->once())
+            ->method('sendAsyncRequest')
+            ->willReturn(new HttpFullfilledPromise(new Response(429, ['Retry-After' => '60'])));
+
+        $transport = new HttpTransport(
+            new Options(['dsn' => $dsn]),
+            $this->httpClient,
+            $this->streamFactory,
+            $this->requestFactory,
+            $this->payloadSerializer,
+            $logger
+        );
+
+        // Event should be sent, but the server should reply with a HTTP 429
+        $promise = $transport->send($event);
+
+        try {
+            $promiseResult = $promise->wait();
```

Comment:
```suggestion
            $promiseResult = $promise->wait();
            $this->fail('Promise should throw');
```

Reply:
The asserts right after the `try`/`catch` block will catch the failure, there is no point to fail programmatically here